// @allurereport/core/dist/history.js
import { mkdir, readFile, writeFile } from "node:fs/promises";
import { dirname, resolve } from "node:path";
import { isFileNotFoundError } from "./utils/misc.js";
const createHistoryItems = (testResults) => {
    return testResults
        .filter((tr) => tr.historyId)
        .map(({ id, name, fullName, historyId, status, error: { message, trace } = {}, start, stop, duration, labels }) => {
        return {
            id,
            name,
            fullName,
            status,
            message,
            trace,
            start,
            stop,
            duration,
            labels,
            historyId: historyId,
            reportLinks: [],
        };
    })
        .reduce((acc, item) => {
        acc[item.historyId] = {
            ...item,
            url: "",
        };
        return acc;
    }, {});
};
export const createHistory = (reportUuid, reportName = "Allure Report", testCases, testResults, remoteUrl = "") => {
    const knownTestCaseIds = testCases.map((tc) => tc.id);
    return {
        uuid: reportUuid,
        name: reportName,
        timestamp: new Date().getTime(),
        knownTestCaseIds,
        testResults: createHistoryItems(testResults),
        metrics: {},
        url: remoteUrl,
    };
};
export const writeHistory = async (historyPath, data) => {
    const path = resolve(historyPath);
    const parentDir = dirname(path);
    await mkdir(parentDir, { recursive: true });
    await writeFile(path, `${JSON.stringify(data)}\n`, { encoding: "utf-8", flag: "a+" });
};
export class AllureLocalHistory {
    constructor(historyPath) {
        this.historyPath = historyPath;
    }
    async readHistory() {
        const path = resolve(this.historyPath);
        try {
            return (await readFile(path, { encoding: "utf-8" }))
                .split("\n")
                .filter((line) => line && line.trim() !== "")
                .map((line) => JSON.parse(line));
        }
        catch (e) {
            if (isFileNotFoundError(e)) {
                return [];
            }
            throw e;
        }
    }
    async appendHistory(data) {
        const path = resolve(this.historyPath);
        const parentDir = dirname(path);
        await mkdir(parentDir, { recursive: true });
        await writeFile(path, `${JSON.stringify(data)}\n`, { encoding: "utf-8", flag: "a+" });
    }
}
